src/eva/vision/models/networks/decoders/linear.py

```diff
@@ -0,0 +1,108 @@
+"""Linear based decoder."""
+
+from typing import List, Tuple
+
+import torch
+from torch import nn
+from torch.nn import functional
+
+from eva.vision.models.networks.decoders import decoder
+
+
+class LinearDecoder(decoder.Decoder):
+    """Linear decoder."""
+
+    def __init__(self, layers: nn.Module) -> None:
+        """Initializes the linear based decoder head.
+
+        Here the input nn layers will be applied to the reshaped
+        features (batch_size, patch_embeddings, hidden_size) from
+        the input (batch_size, hidden_size, height, width) and then
+        unwrapped again to (batch_size, n_classes, height, width).
+
+        Args:
+            layers: The linear layers to be used as the decoder head.
+        """
+        super().__init__()
+
+        self._layers = layers
+
+    def _forward_features(self, features: List[torch.Tensor]) -> torch.Tensor:
+        """Forward function for multi-level feature maps to a single one.
+
+        Args:
+            features: List of multi-level image features of shape (batch_size,
+                hidden_size, n_patches_height, n_patches_width).
+
+        Returns:
+            A tensor of shape (batch_size, hidden_size, n_patches_height,
+            n_patches_width) which is feature map of the decoder head.
+        """
+        if not isinstance(features, list) and features[0].ndim != 4:
+            raise ValueError(
+                "Input features should be a list of four (4) dimensional inputs of "
+                "shape (batch_size, hidden_size, n_patches_height, n_patches_width)."
+            )
+
+        return features[-1]
+
+    def _forward_head(self, patch_embeddings: torch.Tensor) -> torch.Tensor:
+        """Forward of the decoder head.
+
+        Here the following transformations will take place:
+        - (batch_size, hidden_size, n_patches_height, n_patches_width)
+        - (batch_size, hidden_size, n_patches_height * n_patches_width)
+        - (batch_size, n_patches_height * n_patches_width, hidden_size)
+        - (batch_size, n_patches_height * n_patches_width, n_classes)
+        - (batch_size, n_classes, n_patches_height, n_patches_width)
+
+        Args:
+            patch_embeddings: The patch embeddings tensor of shape
+                (batch_size, hidden_size, n_patches_height, n_patches_width).
+
+        Returns:
+            The logits as a tensor (batch_size, n_classes, n_patches_height,
+            n_patches_width).
+        """
+        batch_size, _, height, width = patch_embeddings.shape
```

Comment:
Consider using `hidden_size` explicitely instead of `_` placeholder, or is there a benefit of using `_` here?

Reply:
nice done